diff --git a/pkgs/apps/vips/parsec/gcc-pthreads.bldconf b/pkgs/apps/vips/parsec/gcc-pthreads.bldconf
index e3328c5..cb558ab 100644
--- a/pkgs/apps/vips/parsec/gcc-pthreads.bldconf
+++ b/pkgs/apps/vips/parsec/gcc-pthreads.bldconf
@@ -27,5 +27,5 @@ build_env="CXXFLAGS=\"-I${PARSECDIR}/pkgs/libs/glib/inst/${PARSECPLAT}/include -
 build_inplace="FALSE"
 
 # Arguments to pass to the configure script, if it exists
-vips_basic_conf="--disable-shared --without-fftw3 --without-magick --without-liboil --without-lcms --without-OpenEXR --without-pangoft2 --without-tiff --without-jpeg --without-zip --without-png --without-libexif --without-python --without-x --without-perl --without-v4l"
+vips_basic_conf="--disable-shared --without-fftw3 --without-magick --without-liboil --without-lcms --without-OpenEXR --without-pangoft2 --without-tiff --without-jpeg --without-zip --without-png --without-libexif --without-python --without-x --without-perl --without-v4l --without-cimg"
 build_conf="${vips_basic_conf} --enable-threads"
