diff -Naur gcc-4.2-src-orig/libgomp/config/posix/bar.c gcc-4.2-src-modified/libgomp/config/posix/bar.c
--- gcc-4.2-src-orig/libgomp/config/posix/bar.c	2013-01-17 17:45:24.804248297 -0500
+++ gcc-4.2-src-modified/libgomp/config/posix/bar.c	2013-01-22 15:08:29.720831157 -0500
@@ -31,6 +31,7 @@
    POSIX pthread_barrier_t won't work.  */
 
 #include "libgomp.h"
+#include "tern/user.h"
 
 
 void
@@ -44,6 +45,7 @@
   gomp_sem_init (&bar->sem2, 0);
   bar->total = count;
   bar->arrived = 0;
+  soba_init((long)bar, count, 100);
 }
 
 void
@@ -59,6 +61,7 @@
 #endif
   gomp_sem_destroy (&bar->sem1);
   gomp_sem_destroy (&bar->sem2);
+  soba_destroy((long)bar);
 }
 
 void
@@ -108,4 +111,6 @@
 gomp_barrier_wait (gomp_barrier_t *barrier)
 {
   gomp_barrier_wait_end (barrier, gomp_barrier_wait_start (barrier));
+  soba_wait((long)barrier);
+  
 }
