diff -ruN dkftpbench-0.45/CHECK.cc dkftpbench-0.45-2/CHECK.cc
--- dkftpbench-0.45/CHECK.cc	2002-07-30 13:10:08.000000000 -0700
+++ dkftpbench-0.45-2/CHECK.cc	2014-11-15 20:58:40.198237989 -0800
@@ -18,7 +18,7 @@
 }
 
 /* Check two integers to make sure they are not equal. */
-void checkne(int d, int e, const char* file, int line)
+void checkne(int d, long e, const char* file, int line)
 {
 	if (d == e) {
 		printf("check: %d == %d at line %d file %s\n", d, e, line, file);
diff -ruN dkftpbench-0.45/CHECK.h dkftpbench-0.45-2/CHECK.h
--- dkftpbench-0.45/CHECK.h	2002-01-15 22:21:57.000000000 -0800
+++ dkftpbench-0.45-2/CHECK.h	2014-11-17 17:58:15.060688664 -0800
@@ -15,8 +15,8 @@
 void check(int d, int e, const char * file, int line);
 
 /** Check two integers to make sure they are not equal. */
-#define CHECKNE(d, e) checkne((int)(d), (int)(e), __FILE__,  __LINE__)
-void checkne(int d, int e, const char * file, int line);
+#define CHECKNE(d, e) checkne((int)(d), (long)(e), __FILE__,  __LINE__)
+void checkne(int d, long e, const char * file, int line);
 
 #ifdef __cplusplus
 }
diff -ruN dkftpbench-0.45/dprint.h dkftpbench-0.45-2/dprint.h
--- dkftpbench-0.45/dprint.h	2002-01-15 22:21:58.000000000 -0800
+++ dkftpbench-0.45-2/dprint.h	2014-11-16 12:27:11.213092946 -0800
@@ -24,7 +24,7 @@
 #ifndef dprint_h
 #define dprint_h
 
-#include <stdio.h>
+#include <cstdio>
 
 #ifdef USE_DPRINT
 
diff -ruN dkftpbench-0.45/Platoon.h dkftpbench-0.45-2/Platoon.h
--- dkftpbench-0.45/Platoon.h	2002-02-12 21:06:01.000000000 -0800
+++ dkftpbench-0.45-2/Platoon.h	2014-11-16 12:30:44.361097055 -0800
@@ -1,7 +1,10 @@
 #ifndef Platoon_H
 #define Platoon_H
 #include "robouser.h"
-#include <list.h>
+#include <list>
+#include <cstring>
+#include <cstdlib>
+#include <assert.h>
 
 /* FIXME: 'reasonable' upper bound on number of users CPU can handle */
 #define Platoon_MAXUSERS (1 << 16)
@@ -150,7 +153,7 @@
 	robouser_t *m_users[Platoon_MAXUSERS];
 
 	/** List of dead robousers waiting to be reaped */
-	list<robouser_t *> m_deadlist;
+	std::list<robouser_t *> m_deadlist;
 
 	/** Health: count of bytes fetched so far from all files */
 	size_t m_bytesFetched;
diff -ruN dkftpbench-0.45/robouser.h dkftpbench-0.45-2/robouser.h
--- dkftpbench-0.45/robouser.h	2002-07-04 14:30:24.000000000 -0700
+++ dkftpbench-0.45-2/robouser.h	2014-11-16 12:18:03.637082391 -0800
@@ -23,8 +23,8 @@
 #include "eclock.h"					/* for clock_t - should be eclock_t */
 #include "ftp_client_pipe.h"
 #include "Sked.h"
-#include <stddef.h>
-#include <stdio.h>
+#include <cstddef>
+#include <cstdio>
 
 class Platoon;
 /**--------------------------------------------------------------------------
