diff -ruN pbzip2-1.1.5/Makefile pbzip2-1.1.5-modified/Makefile
--- pbzip2-1.1.5/Makefile	2012-12-25 00:33:45.803045470 -0500
+++ pbzip2-1.1.5-modified/Makefile	2012-12-25 00:14:33.335086704 -0500
@@ -39,6 +39,7 @@
 # External libraries
 LDFLAGS = -lbz2
 LDFLAGS += -lpthread
+LDFLAGS += -I$(XTERN_ROOT)/include -L$(XTERN_ROOT)/dync_hook -Wl,--rpath,$(XTERN_ROOT)/dync_hook -lxtern-annot
 
 # Where you want pbzip2 installed when you do 'make install'
 PREFIX = /usr
diff -ruN pbzip2-1.1.5/pbzip2.cpp pbzip2-1.1.5-modified/pbzip2.cpp
--- pbzip2-1.1.5/pbzip2.cpp	2012-12-25 00:33:42.339045589 -0500
+++ pbzip2-1.1.5-modified/pbzip2.cpp	2012-12-25 00:35:16.314823000 -0500
@@ -279,6 +279,8 @@
 #include <limits.h>
 }
 
+#include <tern/user.h>
+
 
 //
 // GLOBALS
@@ -2644,6 +2646,8 @@
 			return (NULL);
 		}
 
+		soba_wait(0);
+		
 		// compress the memory buffer (blocksize=9*100k, verbose=0, worklevel=30)
 		ret = BZ2_bzBuffToBuffCompress(CompressedData, &outSize,
 				fileData->buf, fileData->bufSize, BWTblockSize, Verbosity, 30);
@@ -3432,6 +3436,8 @@
 	numCPUtotal = numCPU;
 	#endif
 
+	soba_init(0, numCPU, 100);
+
 	// parse command line switches
 	for (i=1; (int)i < argc; i++)
 	{
