--- phoenix-2.0.0-orig/tests/word_count/word_count.c	2013-02-24 15:27:51.826290728 -0500
+++ phoenix-2.0.0/tests/word_count/word_count.c	2013-02-24 23:14:58.365320447 -0500
@@ -41,6 +41,7 @@
 #include "map_reduce.h"
 #include "stddefines.h"
 #include "sort.h"
+#include "tern/user.h"
 
 #define DEFAULT_DISP_NUM 10
 
@@ -148,6 +149,7 @@
  */
 void wordcount_map(map_args_t *args) 
 {
+    soba_wait(0);
     char *curr_start, curr_ltr;
     int state = NOT_IN_WORD;
     int i;
@@ -199,6 +201,7 @@
  */
 void wordcount_reduce(void *key_in, iterator_t *itr)
 {
+    soba_wait(1);
     char *key = (char *)key_in;
     void *val;
     intptr_t sum = 0;
