diff --git a/include/tern/runtime/scheduler.h b/include/tern/runtime/scheduler.h
index c359110..de90faf 100644
--- a/include/tern/runtime/scheduler.h
+++ b/include/tern/runtime/scheduler.h
@@ -229,6 +229,7 @@ struct Serializer: public TidMap {
 
   FILE *logger;
   unsigned* turnCount; // number of turns so far
+  unsigned* turnCount_read; // number of turns so far
 };
 
 
diff --git a/lib/runtime/scheduler.cpp b/lib/runtime/scheduler.cpp
index d98fd65..4ba54b1 100644
--- a/lib/runtime/scheduler.cpp
+++ b/lib/runtime/scheduler.cpp
@@ -45,6 +45,9 @@ Serializer::~Serializer()
   if(turnCount!=NULL){
 	  shmdt((void*)turnCount);
   }
+  if(turnCount_read!=NULL){
+	  shmdt((void*)turnCount_read);
+  }
   if (options::log_sync)
     fclose(logger);
 }
@@ -52,6 +55,7 @@ Serializer::~Serializer()
 Serializer::Serializer(): 
   TidMap(pthread_self()) 
 {    
+  turnCount_read = NULL;
   int shmid;
   if ((shmid = shmget(12345678,8,IPC_CREAT|0666))>=0){
 	  if ((turnCount = (unsigned*)shmat(shmid,0,0))!= (unsigned*)-1 ) {
@@ -64,6 +68,17 @@ Serializer::Serializer():
 	  fprintf(stderr,"you cannot create or open this shared memory\n");
 	  turnCount = NULL;
   }
+  if ((shmid = shmget(87654321,8,IPC_CREAT|0666))>=0){
+	  if ((turnCount_read = (unsigned*)shmat(shmid,0,0))!= (unsigned*)-1 ) {
+		  *turnCount_read = 0;
+	  }else{
+		  fprintf(stderr,"you cannot attach to this shared memory\n");
+		  turnCount_read = NULL;
+	  }
+  }else{
+	  fprintf(stderr,"you cannot create or open this shared memory\n");
+	  turnCount_read = NULL;
+  }
   if (options::log_sync) {
     mkdir(options::output_dir.c_str(), 0777);
     std::string logPath = options::output_dir + "/serializer.log";
@@ -74,6 +89,8 @@ Serializer::Serializer():
 
 unsigned Serializer::incTurnCount(void) { 
   int ret = (*turnCount)++;  
+//  if(NULL!=turnCount_read && (*turnCount_read)!=0)
+//      fprintf(stderr, "%u \n",(*turnCount_read));
   //fprintf(stderr,"the turnCountNow is %u\n",*turnCount);
   if (options::log_sync)
     fprintf(logger, "%d %d\n", (int) self(), ret);
