diff -rupN pfscan-1.0/pfscan.c pfscan-1.0-modified/pfscan.c
--- pfscan-1.0/pfscan.c	2002-03-25 18:13:45.000000000 -0500
+++ pfscan-1.0-modified/pfscan.c	2013-02-19 17:10:43.086741420 -0500
@@ -538,7 +538,8 @@ main(int argc,
     pthread_mutex_init(&print_lock, NULL);
     pthread_mutex_init(&aworker_lock, NULL);
     pthread_mutex_init(&matches_lock, NULL);
-
+    pthread_cond_init(&aworker_cv, NULL);
+    
     for (i = 1; i < argc && argv[i][0] == '-'; i++)
 	for (j = 1; j > 0 && argv[i][j]; ++j)
 	    switch (argv[i][j])
