diff -Naur src_orig/adaptive.C src/adaptive.C
--- parsec-3.0/ext/splash2x/apps/volrend/src_orig/adaptive.C	2013-02-11 18:38:52.368352080 -0500
+++ parsec-3.0/ext/splash2x/apps/volrend/src/adaptive.C	2013-02-11 22:43:21.391844176 -0500
@@ -94,7 +94,14 @@
     LOCK(Global->CountLock);
     Global->Counter--;
     UNLOCK(Global->CountLock);
-    while (Global->Counter);
+    if (!Global->Counter)
+        SETPAUSE(Global->pause)
+    while (Global->Counter) {
+        WAITPAUSE(Global->pause)
+        CLEARPAUSE(Global->pause)
+        if (!Global->Counter)
+            SETPAUSE(Global->pause)
+    }
 
     Ray_Trace_Adaptively(my_node);
 
@@ -135,7 +142,14 @@
     LOCK(Global->CountLock);
     Global->Counter--;
     UNLOCK(Global->CountLock);
-    while (Global->Counter);
+    if (!Global->Counter)
+        SETPAUSE(Global->pause)
+    while (Global->Counter) {
+        WAITPAUSE(Global->pause)
+        CLEARPAUSE(Global->pause)
+        if (!Global->Counter)
+            SETPAUSE(Global->pause)
+    }
 
     Ray_Trace_Non_Adaptively(my_node);
 
@@ -327,7 +339,7 @@
 
 void Ray_Trace_Non_Adaptively(long my_node)
 {
-  long outx,outy,xindex,yindex;
+  long outx,outy,xindex,yindex,done,i;
   float foutx,fouty;
   PIXEL *pixel_address;
 
@@ -372,11 +384,25 @@
       ALOCK(Global->QLock,num_nodes);
       Global->Queue[num_nodes][0]--;
       AULOCK(Global->QLock,num_nodes);
+      SETPAUSE(Global->check)
     }
     local_node = (local_node+1)%num_nodes;
+    done = 0;
     while (Global->Queue[local_node][0] >= lnum_blocks &&
-	   Global->Queue[num_nodes][0] > 0)
-      local_node = (local_node+1)%num_nodes;
+	  Global->Queue[num_nodes][0] > 0 ) {
+      for (i=0;i<num_nodes && !done;++i) {
+        local_node = (local_node+1)%num_nodes;
+        if (Global->Queue[local_node][0] >= lnum_blocks &&
+          Global->Queue[num_nodes][0] > 0 )
+          done = 1;
+      }
+      if (done) {
+        break;
+      }
+      WAITPAUSE(Global->check)
+      CLEARPAUSE(Global->check)
+    }
+    SETPAUSE(Global->check)
   }
 }
 
diff -Naur src_orig/anl.H src/anl.H
--- parsec-3.0/ext/splash2x/apps/volrend/src_orig/anl.H	2013-02-11 18:38:52.364352094 -0500
+++ parsec-3.0/ext/splash2x/apps/volrend/src/anl.H	2013-02-11 22:11:22.731910606 -0500
@@ -31,6 +31,8 @@
   LOCKDEC(IndexLock)
   LOCKDEC(CountLock)
   ALOCKDEC(QLock,MAX_NUMPROC+1)
+  PAUSEDEC(pause)
+  PAUSEDEC(check)
   };
 
 
diff -Naur src_orig/main.C src/main.C
--- parsec-3.0/ext/splash2x/apps/volrend/src_orig/main.C	2013-02-11 18:38:52.364352094 -0500
+++ parsec-3.0/ext/splash2x/apps/volrend/src/main.C	2013-02-11 22:11:22.731910606 -0500
@@ -138,6 +138,8 @@
   LOCKINIT(Global->IndexLock);
   LOCKINIT(Global->CountLock);
   ALOCKINIT(Global->QLock,MAX_NUMPROC+1);
+  PAUSEINIT(Global->pause);
+  PAUSEINIT(Global->check);
 
   /* load dataset from file to each node */
 #ifndef RENDER_ONLY
