--- gcc-4.5.4/libgomp/team.c	2009-05-20 16:54:45.000000000 -0400
+++ gcc-4.5.4-with-detach/libgomp/team.c	2013-03-18 17:58:35.091992182 -0400
@@ -28,6 +28,7 @@
 #include "libgomp.h"
 #include <stdlib.h>
 #include <string.h>
+#include "tern/user.h"
 
 /* This attribute contains PTHREAD_CREATE_DETACHED.  */
 pthread_attr_t gomp_thread_attr;
@@ -59,7 +60,7 @@
 
 /* This function is a pthread_create entry point.  This contains the idle
    loop in which a thread waits to be called up to become part of a team.  */
-
+#include <stdio.h>
 static void *
 gomp_thread_start (void *xdata)
 {
@@ -113,10 +114,15 @@
 	  struct gomp_task *task = thr->task;
 
 	  local_fn (local_data);
+	//fprintf(stderr, "before    wait1 self %u...\n", (unsigned)pthread_self());
 	  gomp_team_barrier_wait (&team->barrier);
+	//fprintf(stderr, "before wait2 self %u...\n", (unsigned)pthread_self());
 	  gomp_finish_task (task);
 
+	//fprintf(stderr, "before       wait3 self %u...\n", (unsigned)pthread_self());
+        tern_detach();
 	  gomp_barrier_wait (&pool->threads_dock);
+	//fprintf(stderr, "before wait4 self %u...\n", (unsigned)pthread_self());
 
 	  local_fn = thr->fn;
 	  local_data = thr->data;
