diff -ruN httpd-2.2.11/server/mpm/worker/worker.c httpd-2.2.11-modified/server/mpm/worker/worker.c
--- httpd-2.2.11/server/mpm/worker/worker.c	2015-08-29 16:42:14.835324898 +0800
+++ httpd-2.2.11-modified/server/mpm/worker/worker.c	2015-08-30 02:27:54.031331042 +0800
@@ -592,6 +592,7 @@
 
 static void *listener_thread(apr_thread_t *thd, void * dummy)
 {
+    tern_init_affinity();
     proc_info * ti = dummy;
     int process_slot = ti->pid;
     apr_pool_t *tpool = apr_thread_pool_get(thd);
@@ -818,6 +819,7 @@
  */
 static void * APR_THREAD_FUNC worker_thread(apr_thread_t *thd, void * dummy)
 {
+    tern_init_affinity();
     proc_info * ti = dummy;
     int process_slot = ti->pid;
     int thread_slot = ti->tid;
